File: python/twelve-days/twelve_days.py
class TwelveDays:
    CARDINALS = {
        1: 'first',
        2: 'second',
        3: 'third',
        4: 'fourth',
        5: 'fifth',
        6: 'sixth',
        7: 'seventh',
        8: 'eighth',
        9: 'ninth',
        10: 'tenth',
        11: 'eleventh',
        12: 'twelfth'
    }

    PHRASES = {
        2: 'two Turtle Doves',
        3: 'three French Hens',
        4: 'four Calling Birds',
        5: 'five Gold Rings',
        6: 'six Geese-a-Laying',
        7: 'seven Swans-a-Swimming',
        8: 'eight Maids-a-Milking',
        9: 'nine Ladies Dancing',
        10: 'ten Lords-a-Leaping',
        11: 'eleven Pipers Piping',
        12: 'twelve Drummers Drumming'
    }

    @classmethod
    def verses(cls, start, stop):
        return "\n".join([cls.verse(i) for i in range(start, stop + 1)]) + "\n"

    @classmethod
    def verse(cls, verse_num):
        return ", ".join(filter(None, [cls.head(verse_num), cls.mid(verse_num),
                                       cls.tail(verse_num)]))

    @classmethod
    def head(cls, verse_num):
        return ("On the %(cardinality)s day of Christmas my true love gave to "
                "me" % ({"cardinality": cls.CARDINALS[verse_num]}))

    @staticmethod
    def tail(verse_num):
        if verse_num == 1:
            return "a Partridge in a Pear Tree.\n"
        return "and a Partridge in a Pear Tree.\n"

    @classmethod
    def mid(cls, verse_num):
        if verse_num != 1:
            return ", ".join([cls.PHRASES[i] for i in range(verse_num, 1, -1)])


def verse(verse_num):
    return TwelveDays.verse(verse_num)


def verses(start, stop):
    return TwelveDays.verses(start, stop)


def sing():
    return TwelveDays.verses(1, 12)

File: java/twelve-days/src/main/java/TwelveDays.java
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class TwelveDays {

  private int FIRST_VERSE = 1;
  private int LAST_VERSE = 12;
  private Map<Integer, String> NUMBER_TO_ORDINAL = numberToOrdinal();
  private Map<Integer, String> NUMBER_TO_GIFT = numberToGift();

  String sing() {
    return verses(FIRST_VERSE, LAST_VERSE);
  }

  String verses(int startVerse, int endVerse) {
    return IntStream.rangeClosed(startVerse, endVerse)
            .mapToObj(currentVerse -> verse(currentVerse))
            .collect(Collectors.joining("\n"));
  }

  String verse(int verseNumber) {
    return prefixForVerse(verseNumber) + giftsForVerse(verseNumber);
  }

  String prefixForVerse(int verseNumber) {
    return String.format("On the %s day of Christmas my true love gave to me, ", ordinal(verseNumber));
  }

  String giftsForVerse(int verseNumber) {
    StringBuilder giftsForVerse = new StringBuilder();
    for (int currentVerse = verseNumber; currentVerse > 0; currentVerse -= 1) {
      if (verseNumber > 1 && currentVerse == 1) {
        giftsForVerse.append("and ");
      }
      giftsForVerse.append(giftsForNumber(currentVerse));
    }
    giftsForVerse.append("\n");
    return giftsForVerse.toString();
  }


  String ordinal(int number) {
    return NUMBER_TO_ORDINAL.get(number);
  }

  String giftsForNumber(int number) {
    if (number > 1) {
      return NUMBER_TO_GIFT.get(number) + ", ";
    }
    return NUMBER_TO_GIFT.get(number);
  }

  private Map<Integer, String> numberToGift() {
    Map<Integer, String> numberToGift = new HashMap<Integer, String>();
    numberToGift.put(1, "a Partridge in a Pear Tree.");
    numberToGift.put(2, "two Turtle Doves");
    numberToGift.put(3, "three French Hens");
    numberToGift.put(4, "four Calling Birds");
    numberToGift.put(5, "five Gold Rings");
    numberToGift.put(6, "six Geese-a-Laying");
    numberToGift.put(7, "seven Swans-a-Swimming");
    numberToGift.put(8, "eight Maids-a-Milking");
    numberToGift.put(9, "nine Ladies Dancing");
    numberToGift.put(10, "ten Lords-a-Leaping");
    numberToGift.put(11, "eleven Pipers Piping");
    numberToGift.put(12, "twelve Drummers Drumming");
    return numberToGift;
  }

  private Map<Integer, String> numberToOrdinal() {
    Map<Integer, String> numberToOrdinal = new HashMap<Integer, String>();
    numberToOrdinal.put(1, "first");
    numberToOrdinal.put(2, "second");
    numberToOrdinal.put(3, "third");
    numberToOrdinal.put(4, "fourth");
    numberToOrdinal.put(5, "fifth");
    numberToOrdinal.put(6, "sixth");
    numberToOrdinal.put(7, "seventh");
    numberToOrdinal.put(8, "eighth");
    numberToOrdinal.put(9, "ninth");
    numberToOrdinal.put(10, "tenth");
    numberToOrdinal.put(11, "eleventh");
    numberToOrdinal.put(12, "twelfth");
    return numberToOrdinal;
  }

}
